### PR TITLE
Site Editor: fix template for page-on-front option

### LINKED ISSUE #66687: Site Editor Uses Incorrect Template
### Description

Since #66579, the Site Editor has changed which template it uses.

In current trunk, the Site Editor uses the template from `/wp/v2/templates/lookup&slug=front-page&_locale=user`.

In Gutenberg 19.5.1, the Site Editor uses the template from `/wp/v2/templates/lookup&slug=page-{$page_slug}&_locale=user`.

Depending on the theme, the templates returned by those two requests can be very different.

### Step-by-step reproduction instructions

1. Create a page.
2. Set that page to be your site's homepage (`page_on_front`).
3. Switch theme to Twenty Twenty-Four.
4. Go to `wp-admin/site-editor.php?canvas=edit`
   * In Gutenberg 19.5.1, you'll see the title and contents of the page you created in step 1. In the Page settings sidebar, note that it's using the "Pages" template (or whatever template you gave it.)
   * In current trunk, you'll see the Twentytwenty-Four default homepage. In the Page settings sidebar, note that it's using the "Blog Home" template.

### Screenshots, screen recording, code snippet

Reading Settings:
Reading Settings

|19.5.1|Trunk|
|-----|------|
|Gutenberg 19.5.1|Current Trunk|

### Environment info

WordPress: 6.8-alpha-59335
Gutenberg: 49c3bf9b8791546231826feeff8681ed3d9ab153 (behavior changed in #66579)
Theme: Twenty Twenty-Four

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [ ] Block
- [ ] Classic
- [x] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

I removed a bit too much logic in #66579. I thought the `/lookup` REST API request could handle fetching the front-page template when the page-on-front option is set, but it can't. Ideally we should fix the endpoint, but in the meantime, I'll restore the previous logic (and add more inline comments).

## Why?

Fixes #66687.

## How?

Restores logic from #66579, closer to where it's needed and with more comments.

When a front page ID is set, the correct front-page hierarchy is first the `front-page` template, then the specific page template for the front page ID. This was previously manually checked in JS. It looked specifically for the front page template and if it doesn't exist, continue with the remaining logic to find the page template.

## Testing Instructions

Set a page to be the front page in Reading settings. The front-end's front page should have changed. Check that it's the same in the site editor.

### CODE REVIEW

File: packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js [hunk 1]
```diff
@@ -96,7 +96,32 @@ function useResolveEditedEntityAndContext( { postId, postType } ) {
...
+						if ( id ) {
+							return id;
+						}
+
+						// If not front page template is found, continue with
+						// the logic below (fetching the page template).
+					} else {
+						// Still resolving `templates`.
+						return undefined;
+					}
```
Comment (ellatrix) [inline packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js:124]: See https://github.com/WordPress/gutenberg/pull/66579/files#diff-90ac1a8edea4ffca3eb4acd2bc2aeed05d54392fc843eb1f253403ae66771dceL59-L66, which has the same logic, but spread over three different pieces: two in the first selector, and one in the second. I put them all together here.

**Review by draganescu — APPROVED** [review]
This solves the problem in my testing.

**Comment by annezazu** [comment]
Thanks for fixing this so quickly!

**Comment by ellatrix** [comment]
Thanks for not reverting my PR 😄

**Pull request merged**